### DISCUSSION TITLE
Timer Support...

### DISCUSSION THREAD

**Discussion by jeffc-dev** [skiptools/skip#114] [Ideas]
If the `Unresolved reference: Timer` errors are anything to go by, it doesn't look like Skip currently supports `Timer` in any of its various forms.

It would be helpful to see this implemented at some point. There are an awful lot of scenarios that really depend upon having the ability to run code on a regular schedule.

**Comment by marcprux** [comment]
It is true we don't yet have low-level support to `Foundation.Timer`. There isn't a perfect analogue on Android to the `RunLoop` that Foundation uses for scheduling. We might be able to put something together if it is a priority for you – just let us know.

But in general, for this kind of thing we recommend using the more modern `Task` support in SwiftUI for scheduling tasks. For example, to perform some periodic updates from your user interface, your top-level content view might do something like:

```swift
struct ContentView : View {
    var body: some View {
        TabView {
            TabView1()
            TabView2()
            TabView3()
        }
        .task(priority: .background) {
            while !Task.isCancelled {
                do {
                    try await Task.sleep(nanoseconds: 1_000_000_000) // 1 second
                    logger.log("executing periodic task at: \(Date.now)")
                    try await viewModel.performUpdates() // or whatever…
                } catch {
                    logger.error("error executing periodic task: \(error)")
                }
            }
        }
    }
}
```

Please let us know if that is sufficient for your needs. If not, I'm sure we can put together a solution.

**Comment by jeffc-dev** [comment]
I think I might be able to use this as a workaround of sorts, but the framework I'm creating (a web service API client that keeps track of the last time updates occurred and periodically performs them as necessary) isn't inherently SwiftUI-specific. Ideally it could handle the periodic checks itself.

I think I could probably use the above as a workaround to get this functional for my needs, though, so as long as full Timer support is on the roadmap _somewhere_ there's no need to develop a prioritized solution for me.

My first intended use will tie into SwiftUI so I think I can use the above method above kick off its _do-I-need-to-do-an-update_ logic. I imagine others will run across this limitation, but I also know that you still have a lot on your plate.

Thanks!

**Reply by aabewhite** [reply]
@jeffc-dev In case we get to this soon, can you please list what Timer APIs you need?

**Comment by aabewhite** [comment]
SkipFoundation now includes basic Timer support. SkipModel and SkipUI have also been updated to allow you to use Timer.publish(...).autoconnect() and View.onReceive(Publisher)

Let us know if you run into any problems!

**Reply by jeffc-dev** [reply]
Thanks! I took a look at the commit and I believe you've covered the bases for what I need.

I will be unable to take a closer look for a few days, but I'll give it a look when I get a chance to revisit my implementation and let you know if I run into any problems.

Thanks again!

**Reply by jeffc-dev** [reply]
I've done a little preliminary testing with the new Timer support (mainly `Timer.scheduledTimer(withTimeInterval:repeats:block:`) and it seems to work well.

Thanks again!
